crates/iota-data-ingestion-core/src/lib.rs

```diff
@@ -51,21 +50,25 @@ pub use worker_pool::WorkerPool;
 #[async_trait]
 pub trait Worker: Send + Sync {
     type Error: Debug + Display;
+    type Message: Send + Sync;
 
-    async fn process_checkpoint(&self, checkpoint: &CheckpointData) -> Result<(), Self::Error>;
-    /// Optional method. Allows controlling when workflow progress is updated in
-    /// the progress store. For instance, some pipelines may benefit from
-    /// aggregating checkpoints, thus skipping the saving of updates for
-    /// intermediate checkpoints. The default implementation is to update
-    /// the progress store for every processed checkpoint.
-    async fn save_progress(
+    async fn process_checkpoint(
         &self,
-        sequence_number: CheckpointSequenceNumber,
-    ) -> Option<CheckpointSequenceNumber> {
-        Some(sequence_number)
-    }
+        checkpoint: &CheckpointData,
+    ) -> Result<Self::Message, Self::Error>;
 
     fn preprocess_hook(&self, _: &CheckpointData) -> Result<(), Self::Error> {
         Ok(())
     }
 }
+
+#[async_trait]
+pub trait Reducer<R: Send + Sync>: Send + Sync {
```

Comment:
Nit: On account of https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-bounds
```suggestion
pub trait Reducer<R>: Send + Sync {
```